### DISCUSSION TITLE
Read/Write WorkMode, MinSoC/MaxSoC and so on for H3-Systems

### DISCUSSION THREAD

**Discussion by abuesen** [nathanmarlor/foxess_modbus#503] [Ideas]
Started with my question in issue #498 @canton7 and me figured out that there are holding-register readable and writeable on the H3 that assumable were not before. 

System: H3-12.0-E - AUX
Firmware Master : 1.83, Slave : 1.03, Manager : 1.66

tested 41000-41017 with `modpoll -m tcp -a 3 -1 -0 -r 41000 192.168.x.x`
Success: 41000, 41007-41011. The answers seem plausible compared to "SinglePhase" and the FoxESS-Cloud
Fail: 41001-41006, 41012, 41013, 41015 (illegal address)
Strange: 41014, 41016 answers not plausible to "SinglePhase" documentation. 41015 is wrong value and 41016 gives e.g. '3789'

First I used "write register" service to set values, it worked. To test I then altered code in my local foxess_modbus and was able to build entities for _select.fox_work_mode_, _sensor.fox_min_soc_on_grid_ and _number.fox_min_soc_on_grid_

Because of an error while reading multible registers after 41000 I had to invalidate some register, too.

I don't know how to pull, push and so on, so I give my changes here. Please put it in a beta the right way, I will test it.
If you want me to do tests on other register let me know. 

[entitiy_descriptions.py line1468ff](https://github.com/nathanmarlor/foxess_modbus/blob/v1.8.2b1/custom_components/foxess_modbus/entities/entity_descriptions.py#L1468)
```
    ModbusSelectDescription(
        key="work_mode",
        address=[
            ModbusAddressSpec(models=[*H1_SET, KH], input=41000),
            ModbusAddressSpec(models=[*H3_SET], holding=41000) # new
        ],
        name="Work Mode",
        options_map={0: "Self Use", 1: "Feed-in First", 2: "Back-up"},
    ),
```

[entitiy_descriptions.py line1570ff](https://github.com/nathanmarlor/foxess_modbus/blob/v1.8.2b1/custom_components/foxess_modbus/entities/entity_descriptions.py#L1570)
```
    ModbusSensorDescription(
        key="min_soc_on_grid",
        addresses=[
            ModbusAddressesSpec(models=[*H1_SET, KH], input=[41011]),
            ModbusAddressesSpec(models=[*H3_SET], holding=[41011]) # new
        ],
        name="Min SoC (On Grid)",
        device_class=SensorDeviceClass.BATTERY,
        state_class=SensorStateClass.MEASUREMENT,
        native_unit_of_measurement="%",
        icon="mdi:battery-arrow-down",
        validate=[Range(0, 100)],
    ),
    ModbusNumberDescription(
        key="min_soc_on_grid",
        address=[
            ModbusAddressSpec(models=[*H1_SET, KH], input=41010),
            ModbusAddressSpec(models=[*H3_SET], holding=41011) # new
        ],
        name="Min SoC (On Grid)",
        mode=NumberMode.BOX,
        native_min_value=10,
        native_max_value=100,
        native_step=1,
        native_unit_of_measurement="%",
        device_class=NumberDeviceClass.BATTERY,
        icon="mdi:battery-arrow-down",
        validate=[Range(0, 100)],
    ),

```
[inverter_profiles.py line 161ff](https://github.com/nathanmarlor/foxess_modbus/blob/ef05fdd2e2d0a9d3821e5da0d954400eedca519b/custom_components/foxess_modbus/inverter_profiles.py#L161)
```
        InverterModelProfile(H3, r"^H3-")
        .add_connection_type(
            LAN,
            RegisterType.HOLDING,
            invalid_register_ranges=invalid_ranges.H3, # new
        )
        .add_connection_type(
            AUX,
            RegisterType.HOLDING,
            invalid_register_ranges=invalid_ranges.H3, # new
        ),

```

[invalid_ranges.py appended](https://github.com/nathanmarlor/foxess_modbus/blob/v1.8.2b1/custom_components/foxess_modbus/entities/invalid_ranges.py)
`H3 = [(41001, 41002, 41003, 41004, 41005, 41006, 41013, 41015)]`

The errormessage:
```
Logger: custom_components.foxess_modbus.modbus_controller
Source: custom_components/foxess_modbus/modbus_controller.py:221
Integration: FoxESS - Modbus (documentation, issues)
First occurred: 10:59:45 (1 occurrences)
Last logged: 10:59:45

tcp://192.168.x.x:502 3 - 5 failed poll attempts: now not connected. Last error: Error reading registers. Type: RegisterType.HOLDING; start: 41000; count: 12; slave: 3 from tcp://192.168.x.x:502: Exception Response(131, 3, IllegalAddress)
```

![image](https://github.com/nathanmarlor/foxess_modbus/assets/136979340/ac08c661-2038-434f-9b49-c6b6bf86a2d7)

**Comment by abuesen** [comment]
Register 41014 is '1' if set 50Hz and results in _TimeOut_ at 60Hz. Seems like broken or work in progress.

**Comment by canton7** [comment]
Oh amazing, thank you! Let me get those in, and a new beta out for you.

**Reply by canton7** [reply]
Right, I've released v1.8.2b2 (and I'm sure I didn't mess up the release this time!)

* I've taken out the charge period config
* I've added in 41001-41006, 41012-41013, 41015 as invalid registers
* I've kept work_mode, min_soc, min_soc_on_grid, max_soc, max_charge_current and max_discharge_current

You didn't have max_charge_current and max_discharge_current in your list of changes, but you said the registers seemed to work, so I left them in.

**Reply by abuesen** [reply]
Yes, that was my intention. I just wanted a proof of concept and WorkMode is very interesting for me :-)
I'll test.

**Reply by TonyM1958** [reply]
According to my info, 41001-41006 are not invalid registers. You may be getting illegal address on 41001-41006 as this is block of 6 registers for the charge times. On H1, they only work if you write them as a block of 6 although you can read them indiviudally. It could be that, on H3, you need to read and write these registers as a block of 6 for them to work?

**Reply by canton7** [reply]
@TonyM1958 See further down: https://github.com/nathanmarlor/foxess_modbus/discussions/503#discussioncomment-8119999

There doesn't seem to be any way to read them on the H3. Trying to read them as part of a larger block also fails, so I need to tell the integration not to create a block read which spans them. That's what the invalid register ranges are for.

**Reply by TonyM1958** [reply]
OK - long shot, but strategy periods seem to lock out updating a number of registers, includng charge times. I assume read / write of the charge times was tested with strategy periods disabled? As you say, would be nice to get this working but it's not easy when you don't have direct access to an inverter...

**Reply by canton7** [reply]
To be honest, we've got force charge/discharge in the integration now (well, in beta), so charge periods are a bit moot.

**Reply by abuesen** [reply]
> charge times was tested with strategy periods disabled?

@TonyM1958 
What are strategy periods and how to disable? Can't find them in modbus.map.

**Reply by TonyM1958** [reply]
Strategy periods or schedules are not accessible via modbus currently. They are set under work mode using the Fox app or Web site. They require a firmware upgrade before they can be used. This depends on the model of inverter.

**Reply by abuesen** [reply]
Is it another thing than the time-periods? See screenshot. In the web and in the app there are only Work Mode and two time periods, both disabled.
The system should be on the latest firmware. At the moment I have an account type "Enduser". So I can't update myself but I don't mind until now. 

For my Inverter and software-version see my first porst: https://github.com/nathanmarlor/foxess_modbus/discussions/503#discussion-6066099

![grafik](https://github.com/nathanmarlor/foxess_modbus/assets/136979340/9792e2f2-f4a7-477e-8492-49dbee847792)

**Reply by canton7** [reply]
Yep, your manager firmware isn't new enough to support strategy periods, so this won't be the reason.

**Reply by abuesen** [reply]
Ok, thank you. I'll working on that this week ;-)
Never looked for firmware and releasenotes at FoxESS before.

**Comment by abuesen** [comment]
**First try:** Entites show up, no values at all.
Log (1) shows illegal adresses 41007+5. (Don't understand the linenumbers from log.)
With modpoll 41007-41011 give values. 

**Second try**: I put '#' in froint of the lines with 41007, 41008, 41009 what should be near the same as I did before, but no values.
Log (2) complains about a binary sensor. 

But after reloading foxess_modbus 1.8.2b2 the first error does not appear again. The second one stays. That is not what I expected. I'll reboot and revert to 1.8.1 to get a fresh start. Maybe not today.


Log (1):
```
Logger: custom_components.foxess_modbus.modbus_controller

Source: custom_components/foxess_modbus/modbus_controller.py:221
Integration: FoxESS - Modbus ([documentation](https://github.com/nathanmarlor/foxess_modbus), [issues](https://github.com/nathanmarlor/foxess_modbus/issues))
First occurred: 15:08:14 (1 occurrences)
Last logged: 15:08:14
tcp://192.168.100.7:502 3 - 5 failed poll attempts: now not connected. Last error: Error reading registers. Type: RegisterType.HOLDING; start: 41007; count: 5; slave: 3 from tcp://192.168.100.7:502: Exception Response(131, 3, IllegalAddress)
```
Log (2)
[home-assistant_2024-01-13T16-26-19.691Z.log](https://github.com/nathanmarlor/foxess_modbus/files/13928840/home-assistant_2024-01-13T16-26-19.691Z.log)

```
Logger: homeassistant.components.binary_sensor
Source: helpers/entity_platform.py:573
Integration: Binärsensor (documentation, issues)
First occurred: 17:18:30 (1 occurrences)
Last logged: 17:18:30

rpi_power: Error on device update!
Traceback (most recent call last):
  File "/usr/src/homeassistant/homeassistant/helpers/entity_platform.py", line 573, in _async_add_entity
    await entity.async_device_update(warning=False)
  File "/usr/src/homeassistant/homeassistant/helpers/entity.py", line 1214, in async_device_update
    await hass.async_add_executor_job(self.update)
  File "/usr/local/lib/python3.11/concurrent/futures/thread.py", line 58, in run
    result = self.fn(*self.args, **self.kwargs)
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/src/homeassistant/homeassistant/components/rpi_power/binary_sensor.py", line 52, in update
    value = self._under_voltage.get()
            ^^^^^^^^^^^^^^^^^^^^^^^
AttributeError: 'NoneType' object has no attribute 'get'

```

**Reply by abuesen** [reply]
Fine: System reboot helped to get the expected error :-)

**Reply by canton7** [reply]
That rpi_power is nothing to do with us, I think.

> I put '#' in froint of the lines with 41007, 41008, 41009 what should be near the same as I did before, but no values.
Log (2) complains about a binary sensor.

So you commented out the sensors with addresses 41007, 41008 and 41009. Then got an unrelated error from the rpi_power integration. Does foxess_modbus load at all in that case?

> System reboot helped to get the expected error :-)

Which error exactly is the "expected error"?

**Reply by abuesen** [reply]
Expeced meant the system to complain about illegal adresses, like the first try :-)
I agree it has nothing to do with our task, but the comments raised it.

New insights:

1. Reload 1.8.2b2 
2. _System_ reboot 
3. (Illegal register)
4. comment out same lines 
5. HA-reload 
6. (rpi_power error)
7. System reboot 
8. (illegal register 41010 (log attached))
[home-assistant_2024-01-13T16-55-36.558Z.log](https://github.com/nathanmarlor/foxess_modbus/files/13928913/home-assistant_2024-01-13T16-55-36.558Z.log)

So there must be another problem now. Register 41010 worked before.

**Reply by canton7** [reply]
I wonder if the problem is reading them as a block. Can you try `modpoll -m tcp -a 3 -1 -0 -r 41010 -c 2 192.168.x.x`

**Reply by abuesen** [reply]
:Yes: reading 41010 and 41011 as block fails. Single read works.

**Reply by canton7** [reply]
Huh. That's annoying: I'm going to have to build some logic for polling registers individually...

**Reply by abuesen** [reply]
Strange, the rpi_power error persits now. I reverted to 1.8.1 and inserted the lines I did before. System-reboot but still this error.

**Reply by canton7** [reply]
I'm pretty sure that's nothing to do with foxess_modbus

**Reply by abuesen** [reply]
Right. Reason is the comment within the brackets '[...]' in _entity_descriptions.py_. I copied the '# new' from my post and it causess this error.
Now it looks like before 😃

**Reply by canton7** [reply]
Cool! I've released v1.8.2b3 which should be reading registers in the range 41007-41011 one-by-one. Can you give that a try please?

**Comment by abuesen** [comment]
Very nice: 1.8.2b3 works. 🎉 

It starts and gives values for all entities. Now I have to check function.
Did it already for _WorkMode_, _Min SoC on Grid_ and _Min SoC._

Open for _Max SoC_ and _Currents_, I have to  look for other places to see this values.
_Max SoC_ shows 75% what is strange because last days with sun it goes 100%.
Both currents show 26A what is plausible.

**Reply by canton7** [reply]
Awesome!

Max soc only affects charging from the grid - charging from PV ignores this value.

**Reply by canton7** [reply]
If you get the chance, could you have another go at reading those charge period registers (one by one, in a block of 3, a block of 6, a block of 1 and then 2, etc)? I know a lot of people wol be extremely happy if we get H3 charge periods working!

**Reply by abuesen** [reply]
The chance is to use modpoll, isn't it? ;-) 
Need some assitance for the correct calls. Is one by one different to block of 1?

 I did:
* `modpoll -m tcp -a 3 -1 -0 -r 41001 192.168.x.x` with register 41001-41006
All "Illegal Data Address"
* `modpoll -m tcp -a 3 -1 -0 -r 41001 -c 1 192.168.x.x` with '-c n', n=1-6
All "Illegal Data Address"
* `modpoll -m tcp -a 3 -1 -0 -r 41002 -c 2 192.168.x.x`
"Illegal Data Address"
* All the same with '-t 3' like `modpoll -m tcp -a 3 -1 -0 -t 3 -r 41001 192.168.x.x`
All "Time Out"

Anything I missed?


What do they want to do? Assume they want to charge from grid if power is cheap. That is not reachable with the other entities, unfortunaly the workmodes "force charge" and "force dischage" seem outdated. Some other tasks can be done with WorkMode or SoC.

Is it worth to take a close look on "Remote power control" with register 44000-44005? 
Of course one has to do all calculating and control on its own.  That's a huge thing: Understanding the work of the power-register, doing the calculation with all siedeefects and smoothing the control.

**Reply by canton7** [reply]
Aww, those are all the options I had in mind. Thanks for trying!

Yes, you need to use charge periods in order to charge from the grid (which is useful when there's cheap electricity overnight). FoxESS are moving towards schedule periods for force charge / force discharge, but unfortunately those aren't available over modbus (yet?).

**Reply by abuesen** [reply]
It really accepts WorkMode 0-4. No more.

**Reply by canton7** [reply]
If you can work out how "remote power control" works, that might be interesting...

**Reply by abuesen** [reply]
Did starting the "remote power control"-way at beginning, but was very happy to use WorkMode for my task. 😃 And it works very fine since two days! Great integration.

**The start:**
Enabled "Remote Power Control" with 44000=1.  
If the SoC was 100% it stops doing anything. If the SoC was about 45% 

Sun was shining. 
Enabled "Remote Power Control" with 44000=1. 
If the SoC was 100% it stops doing anything. If the SoC was about 45% it stops exporting power to house/grid.
Then I put '2000' or '4000' to 44002 and that was the output from the inverter. Both times feed from PV.

But 44002 and 44003 schould be a pair for controlling active power. Dont know how to use.
And one should be able to set negative values, the service "Write register" did not do (for any reason, mayby foxess)
Last point this behaviour is not the way what the note in the documentation says.

On normal operation reading 44002 and 44003 show values in both registers. Dont remember if there were negetive ones.
In 44004 and 44005 I found nagative values. At the moment the battery is min and my foxess does nothing all values are '0' exept 44003 is '-31131'. 

**What I can do:**
I can try to observe this register while normal operation. maybe it is useful to set them up as sensors to get a history. I think I'm able to do this myself. Learned yesterday ;-) (If I had influx/grafana ready I could corellate the values, but I haven't yet and missing time therefore). 
I can set to control mode an write some values to the register if you can lead me a bit.

**What I can't do** is spend a lot of time to understand the whole thing. I've got only basic knowledg to get my things running. So I cannot work it out completely.

**Reply by canton7** [reply]
Thanks! To send a negative value, do 65536 - the value.

**Reply by abuesen** [reply]
annoing. The _rpi_power_ error seems to raise on any syntaxerror in _entities_description.py_. Did not manage to add a Sensor for 44002 or 44002/44003.

Tried this at the end of the file, _power_control_enable_ works, _rpc_active_power_ doesn't
```
    ModbusSelectDescription(
        key="remote_power_control_enable",
        address=[ModbusAddressSpec(models=H3_SET, holding=44000)],
        name="Remote Power Control",
        options_map={0: "Disabled", 1: "Enabled"},
    ),
    ModbusSensorDescription(
        key="rpc_active_power",
        addresses=[ModbusAddressesSpec(models=H3_SET, holding=[44002])],
        name="RPC active Power",
        device_class=SensorDeviceClass.ENERGY,
        state_class=SensorStateClass.MEASUREMENT,
        native_unit_of_measurement="W",
        icon="mdi:home-lightning-bolt-outline",
        signed=True,
    )
]
```

**Reply by abuesen** [reply]
It seems in normal operation the valus of 44002/44003 and 44004/44005 are not set by the system. The values of 2/3 are still '0' and 4/5 alter somtimes. Cant see any systematic.
(Thank you for bringing signed int16 to my mind again. Stupid question.) 😆 

The PV generates less than 100W at the moment. SoC is about 35%. The Load is feed by PV+battery.
1. _Remote Power Control_ enabled.
No power for the load any more. Battery is charging from PV.
2. Setting 44002 to '500'
Inverter delivers 500W to the load feed from PV+battery.
3. Setting 44002 to '65035' (-501)
Inverter consumes, battery is charing with about 675W what is a bit more than 501+PV, but the last digit is correct, alwas PV-power +1. The difference is alway 40W
4. Setting positve and negative values to 44003 are written and stored
can't see an effect
5. Setting 44004 is possible with negative values only
can't see an effect
6. Setting 44005 is not possible at all
but it shows altering negative values. saw between -8 and -28, but only even numbers.  


Seems nice on the first sight. Hopefully foxess cares for temperature and so on regardless of the external power control.

**Reply by canton7** [reply]
Take a look at #513 - I did some of my own investigation this morning.

**Reply by canton7** [reply]
I would be very surprised if rpi_power is related at all. I think it's randomly failing, and just happens to coincide with other things you're doing.

**Reply by abuesen** [reply]
Ok, you're much more advanced in this work than me of course. :-D 
Feel free to give me tasks for the H3 if I can assist testing. There is no need of remote power control for my own usage.

**Comment by abuesen** [comment]
Read and write both _Current_ and _Max SoC_ works too. Did not test what happens on "wrong" values like more than 100% or high currents. 

So in my opinion this peace works fine for H3. Of course we do not know about other 3-phase inverters and other firmware.

What about catching the "illegal adress" exception instead of the integration does not load?
Maybe it is possible to set the entities to "unavailiable" or set '0' or '999'|'666'

**Reply by abuesen** [reply]
good work, 110% not alowed by _Write register_ and '50A' result in a slave error.

**Comment by peterw10** [comment]
I don't know if its helpful, here's the answer from FoxEss support I got:

"H3 Modbus-Protokoll Einstellung der Zwangsladung
Einstellung des 3-teiligen Registers in Folge, Einstellung 44000 auf 7, Einstellung 44001 für Steuerzeit, Einheit S(Sekunde), Einstellung 44002 für Leistung, Einheit W(Watt)"

**Reply by canton7** [reply]
Writing 7 to 44000 gives "Illegal Data Value exception response!", at least on my H1. The rest of it looks in like with https://github.com/nathanmarlor/foxess_modbus/discussions/513.
